[lua] Mountain worm nm lua

MowFord opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Partner to https://github.com/LandSandBoat/server/pull/5449

[Mountain Worm NM](https://www.bg-wiki.com/ffxi/Mountain_Worm_(NM)) is fairly simple

Has the standard thing that quite a few worms should have, where it only casts if out of melee range. Coded it manually since it's an NM (based on the same logic in `mob_controller.cpp`)

If this is approved can just close my sql PR. 

## Steps to test these changes

`!spawnmob 16798031`, `!gotoid 16798031`, witness fight

TeoTwawki commented on scripts/zones/Uleguerand_Range/mobs/Mountain_Worm_NM.lua:
```diff
@@ -0,0 +1,29 @@
+--  Mob: Mountain Worm NM
+--  https://www.bg-wiki.com/ffxi/Mountain_Worm_(NM)
+-- TODO allow deaggro based on distance (core CMobEntity::CanDeaggro() forces NM and Battlefield mobs to never stop chasing)
+-- TODO confirm if despawns on lost aggro?
```
I can help here!

He doesn't depop, but he will very likely move (just standard random worm movement behavior) to a diff part of the cave system (little jerk...) and be really difficult to find again before he full heals.

MowFord replied:
Ah so he can deaggro (unlike most NM) but doesn't despawn just hard to find again

TeoTwawki replied:
All non battlefield NM on retail can deaggro. It just they have to go a long while without interacting with you, for worms thats easy - be out of range. for other NMs they chase your \*ss down and slap you :)

MowFord replied:
https://github.com/LandSandBoat/server/blob/f0d23cacf13df0a8cc91f8a5342b3351b26fd86a/src/map/entities/mobentity.cpp#L359

👀

TeoTwawki replied:
🤦‍♂️

TeoTwawki commented on scripts/zones/Uleguerand_Range/mobs/Mountain_Worm_NM.lua:
```diff
@@ -0,0 +1,29 @@
+end
+
+entity.onMobFight = function(mob, target)
+    if mob:checkDistance(target) > mob:getModelSize() + 3 then
```
kinda surprised the attempt at checking size wasn't baked into distance checks already (it'd be wrong, but its something i expect the mistake to be made in). 

`getModelSize()` really needs either renamed or how we determine size needs reworked for mobs, because the value stored in `m_ModelRadius` is NOT the physical size of the model on screen: its the "melee range" of the monster stored in `mob_family_system` that currently is not related to its actual size. you can set crazy values there and have a tiny little chigoe that is able to melee hit you and be hit in return from 30 yalms away. And that's the value you are adding 3 to and including here.

The mobs actual size is set in a few bits of a mask that have different purposes depending on which type of entity model it is. In the case of this worm, there are 3 single bits that influence how large it appears in the client. These are currently in the 1st byte of the entityFlags value stored in `mob_pools`

tl;dr You should probably add a comment here about how because he shares a family with the normal mountain worm but is a wee bit larger you had to expand this a bit. At least if I'm correct about that being the case anyway. If I'm not, prolly should drop the +3 then.

TeoTwawki replied:
side thought: we could also bake this into a mob mod or a mixin to avoid writing the same code for future mobs, but I suggest crossing that bridge when we get to another mob we want that behavior on.

MowFord replied:
Btw I didn't add this because of his size, I did it because that's the exact logic in core to determine if a mob will move or auto attack

TeoTwawki replied:
so you remade logic related to movement for a mob that _doesn't move._ 🤔 

I know that spot in the core, I've touched it myself. its...less than optimal and fudges on the side of making sure players can actually hit things they are standing in front of despite how fat king behemoths rear end may be. Now I'm thinking again about the mobmod line of thought I was having, maybe that bridge _should_ be crossed sooner rather than later, instead of running a distance check a and mob radius check then attempting to set its casting ability every 240ms.


I think you really wanted "can I hit the player?". I really hope the logic in the **draw in** code that works on that very premise doesn't do the same thing.

_looks_

```cpp
if (currentDistance >= PMob->GetMeleeRange() * 2 && battleutils::DrawIn(PTarget, PMob, PMob->GetMeleeRange() - 0.2f))
```

```cpp
float CBattleEntity::GetMeleeRange() const
{
    return m_ModelRadius + 3.0f;
}
```

_**"Son of a...**_

😭 we really can't win can we. We may as well lua bind `getMeleeRange()` and save some future steps at least, because _some day_ those size bits are going to be correctly used and then melee range will get updated to account for them.

MowFord replied:
If we want that binding I can add it, sure

TeoTwawki replied:
> If we want that binding I can add it, sure

yeah lets do that for now theres bound to be other times we want to know melee range in lua

MowFord replied:
Or perhaps I could remove this logic completely and let [this PR](https://github.com/LandSandBoat/server/pull/5313/files) handle it

TeoTwawki replied:
I'd forgotten about that, last I recall it needed to sort out which worms did and didn't do this behavior. But if you wnat to leave the casting behavior to that, that's find by me as well.

MowFord replied:
I'll remove the mob fight logic altogether and watch for that PR to include this NM. Ezpz

TeoTwawki commented on sql/mob_spawn_points.sql:
```diff
@@ -1101,7 +1101,7 @@ INSERT INTO `mob_spawn_points` VALUES (16798027,'Nightmare_Bats','Nightmare Bats
 INSERT INTO `mob_spawn_points` VALUES (16798029,'Mountain_Worm','Mountain Worm',42,-267.339,-37.646,-141.238,1);
 INSERT INTO `mob_spawn_points` VALUES (16798030,'Mountain_Worm','Mountain Worm',42,-202.226,-30.116,-179.173,0);
-INSERT INTO `mob_spawn_points` VALUES (16798031,'Mountain_Worm','Mountain Worm',48,-269.816,-79.023,18.257,3);
+INSERT INTO `mob_spawn_points` VALUES (16798031,'Mountain_Worm_NM','Mountain Worm',48,-269.816,-79.023,18.257,3);
```
👍

TeoTwawki commented on scripts/zones/Uleguerand_Range/mobs/Mountain_Worm_NM.lua:
```diff
@@ -0,0 +1,24 @@
+local entity = {}
+
+entity.onMobSpawn = function(mob)
+    mob:setMagicCastingEnabled(false)
```
missed a spot when removing the onMobFight part

MowFord replied:
Oops, thank you

MowFord replied:
updated

TeoTwawki commented on scripts/zones/Uleguerand_Range/mobs/Mountain_Worm_NM.lua:
```diff
@@ -0,0 +1,24 @@
+-- Area: Uleguerand Range
+--  Mob: Mountain Worm NM
+--  https://www.bg-wiki.com/ffxi/Mountain_Worm_(NM)
+-- TODO allow deaggro based on distance (core CMobEntity::CanDeaggro() forces NM and Battlefield mobs to never stop chasing)
```
Its technically not the distance (I think) but just how long it goes without being able to target/attack you in any way. which of course would mean you are out of casting range.

MowFord committed:
[lua] Mountain Worm NM

TeoTwawki committed:
Amend comment in Mountain_Worm_NM.lua

I don't want to potentially confuse future contributors: The mob isn't going to instant lose you the second you are out of range, but rather has the normal deaggro mechanism if it cannot interact with you for an extended amount of time.

TeoTwawki reviewed (approved):
I amended your TODO comment
